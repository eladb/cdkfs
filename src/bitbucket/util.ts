

export function reduceRenderObject( obj:Record<string, any> ) {

  const result = Object.keys(obj).reduce( ( inp:Record<string, any>, key:string) => {

    if (obj[ key ] !== undefined) inp[ key ] = obj[ key ];

    return inp;
  }, {} );

  if ( Object.keys( result).length === 0 ) return undefined;

  return result;
}


export function reduceRenderArray( arr:any[]|undefined ) {

  if ( arr === undefined ) return arr;
  else if (arr.length === 0) return undefined;
  return arr;
}