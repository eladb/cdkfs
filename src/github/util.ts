export function secretToString(secretName: string): string {
  return `\${{ secrets.${secretName} }}`;
}

/**
 * Bash if equivalent execution of scripts that is compatible with shx.
 * Execution will fail, if the when or then command fails.
 *
 * @param when The condition to evaluate. Usually a `test` command. This command may be executed twice. Make sure it is non mutating.
 * @param then The script to run if the condition is true.
 * @param otherwise The script to run if the condition is false.
 */
export function shxIf(when: string, then: string, otherwise?: string): string {
  // Connect the condition and the 'then' command with a logical AND.
  // If the condition is true, this will execute the 'then' command.
  // This case is equivalent to TRUE && then()
  // If the condition is false, this will exit with a non-zero status code
  // (indicating an error) without executing the 'then' command.
  // This case is equivalent to FALSE && then()
  const if_clause = `${when} && (${then})`;

  // For the else clause, we need to check the 'when' condition again,
  // because the if clause can also return FALSE if the 'then' command fails.
  // However in this situation we don't want to execute the else clause but report an error.
  const else_clause = `${when} || (${otherwise})`;

  // Try the if_clause first. If it fails try the else clause.
  if (otherwise) {
    return `(${if_clause}) || (${else_clause})`;
  }

  // Just run the if clause
  return if_clause;
}
