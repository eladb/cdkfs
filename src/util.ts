import * as child_process from 'child_process';
import * as path from 'path';
import * as fs from 'fs-extra';
import * as logging from './logging';
// eslint-disable-next-line @typescript-eslint/no-require-imports
const decamelize = require('decamelize');

export function exec(command: string, options?: child_process.ExecSyncOptions) {
  logging.verbose(command);
  return child_process.execSync(command, {
    stdio: ['inherit', process.stderr, 'pipe'],
    ...options,
  });
}

/**
 * Executes `command` and returns its value or undefined if the command failed.
 */
export function execOrUndefined(command: string, options?: child_process.ExecSyncOptions): string | undefined {
  try {
    const value = child_process.execSync(command, { stdio: ['inherit', 'pipe', 'ignore'], ...options }).toString('utf-8').trim();
    if (!value) { return undefined; } // an empty string is the same as undefined
    return value;
  } catch {
    return undefined;
  }
}

export interface WriteFileOptions {
  /**
   * Whether the generated file should be marked as executable.
   *
   * @default false
   */
  executable?: boolean;

  /**
   * Whether the generated file should be readonly.
   *
   * @default false
   */
  readonly?: boolean;
}

export function getFilePermissions(options: WriteFileOptions): string {
  const readonly = options.readonly ?? false;
  const executable = options.executable ?? false;
  if (readonly && executable) {
    return '500';
  } else if (readonly) {
    return '400';
  } else if (executable) {
    return '755';
  } else {
    return '644';
  }
}

export function writeFile(filePath: string, data: any, options: WriteFileOptions = {}) {
  if (fs.existsSync(filePath)) {
    fs.chmodSync(filePath, '600');
  }

  fs.mkdirpSync(path.dirname(filePath));
  fs.writeFileSync(filePath, data);

  fs.chmodSync(filePath, getFilePermissions(options));
}

/**
 * Decamelizes the keys of an object structure, recursing through child objects and arrays.
 * @experimental
 */
export interface DecamelizeRecursivelyOptions {
  /**
   * Max depth to recurse before erroring.
   * @default 10
   */
  maxDepth?: number;

  /**
   * Returns true when a key should be decamelized
   * @default - all keys are decamelized
   */
  shouldDecamelize?: (path: string[], value: any) => boolean;

  /**
   * Separator for decamelizing.
   * @default "_"
   */
  separator?: string;

  /**
   * Current path.
   * @internal
   */
  path?: string[];
}

export function decamelizeKeysRecursively(input: any, opt?: DecamelizeRecursivelyOptions): any {
  const shouldAlwaysDecamelize = () => true;
  const shouldDecamelize = opt?.shouldDecamelize ?? shouldAlwaysDecamelize;
  const separator = opt?.separator ?? '_';
  const path_ = opt?.path ?? [];
  const maxDepth = opt?.maxDepth ?? 10;

  if (path_.length > maxDepth) {
    throw new Error('Decamelled too deeply - check that the input has no circular references');
  }

  if (Array.isArray(input)) {
    return input.map((k, i) => decamelizeKeysRecursively(k, {
      ...opt,
      path: [...path_, i.toString()],
    }));
  }

  if (typeof input === 'object' && input !== null) {
    const mappedObject: Record<string, any> = {};
    for (const [key, value] of Object.entries(input)) {
      const transformedKey = shouldDecamelize([...path_, key], value)
        ? decamelize(key, separator)
        : key;

      mappedObject[transformedKey] = decamelizeKeysRecursively(value, {
        ...opt,
        path: [...path_, key],
      });
    }

    return mappedObject;
  }

  return input;
}

/**
 * Returns false if value is unset or a falsey value, and true otherwise.
 * @param value an environment variable
 */
export function isTruthy(value: string | undefined): boolean {
  return !(value === undefined || ['null', 'undefined', '0', 'false', ''].includes(value.toLocaleLowerCase()));
}

/**
 * Type of a map mapping strings to some arbitrary type
 */
export type Obj<T> = { [key: string]: T };

/**
 * Return whether the given value is an object
 *
 * Even though arrays and instances of classes technically are objects, we
 * usually want to treat them differently, so we return false in those cases.
 */
export function isObject(x: any): x is Obj<any> {
  return x !== null && typeof x === 'object' && !Array.isArray(x)
    && x.constructor.name === 'Object';
}

/**
 * Recursively merge objects together
 *
 * The leftmost object is mutated and returned. Arrays are not merged
 * but overwritten just like scalars.
 *
 * If an object is merged into a non-object, the non-object is lost.
 *
 * `undefined`s will cause a value to be deleted if destructive is enabled.
 */
export function deepMerge(objects: Array<Obj<any> | undefined>, destructive: boolean = false) {
  function mergeOne(target: Obj<any>, source: Obj<any>) {
    for (const key of Object.keys(source)) {
      const value = source[key];

      if (isObject(value)) {
        // if the value at the target is not an object, override it with an
        // object so we can continue the recursion
        if (!isObject(target[key])) {
          target[key] = value;
        }
        mergeOne(target[key], value);

        // if the result of the merge is an empty object, it's because the
        // eventual value we assigned is `undefined`, and there are no
        // sibling concrete values alongside, so we can delete this tree.
        const output = target[key];
        if (isObject(output) && Object.keys(output).length === 0 && destructive) {
          delete target[key];
        }
      } else if (value === undefined && destructive) {
        delete target[key];
      } else if (typeof value !== 'undefined') {
        target[key] = value;
      }
    }
  }

  const others = objects.filter(x => x != null) as Array<Obj<any>>;

  if (others.length === 0) { return {}; }
  const into = others.splice(0, 1)[0];

  others.forEach(other => mergeOne(into, other));
  return into;
}

/*
 * Deduplicate values in a list, returning a new array.
 * @param array list of values
 */
export function dedupArray<T>(array: T[]): T[] {
  return array.filter((val, idx) => array.indexOf(val) === idx);
}

/**
 * Returns a sorted version of `x` or `undefined` if it is an empty array or object.
 */
export function sorted<T>(x: T) {
  if (x == null) { return undefined; }
  if (Array.isArray(x)) {
    if (x.length === 0) { return undefined; }
    return (x as unknown[]).sort();
  } else if (typeof x === 'object') {
    if (Object.keys(x).length === 0) { return undefined; }
    const result: Record<string, unknown> = {};
    for (const [key, value] of Object.entries(x).sort(([l], [r]) => l.localeCompare(r))) {
      result[key] = value;
    }
    return result as T;
  } else {
    return x;
  }
}

export function formatAsPythonModule(name: string) {
  return name.replace(/-/g, '_').replace(/\./g, '_');
}
