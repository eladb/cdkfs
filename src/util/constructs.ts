import type { IConstruct } from "constructs";
import type { Component } from "../component";
import type { Project } from "../project";

export const PROJECT_SYMBOL = Symbol.for("projen.Project");
export const COMPONENT_SYMBOL = Symbol.for("projen.Component");

/**
 * Create a function to find the closest construct matching a predicate
 * @param predicate
 * @returns A function to find the closest construct matching the predicate
 */
export function tryFindClosest<T extends IConstruct>(
  predicate: (x: any) => x is T
): (construct?: IConstruct) => T | undefined {
  return (construct?: IConstruct) =>
    construct?.node.scopes.reverse().find(predicate);
}

/**
 * Create a function to find the closest construct matching a predicate
 * @param predicate
 * @returns A function to find the closest construct matching the predicate
 */
export function findClosestProject(construct: IConstruct): Project {
  if (isComponent(construct)) {
    return construct.project;
  }

  const project = tryFindClosest(isProject)(construct);
  if (!project) {
    throw new Error(
      `${new.target.name} at '${
        construct.node.path
      }' must be created in the scope of a Project, but no Project was found`
    );
  }

  return project;
}

export function isProject(x: any): x is Project {
  return x !== null && typeof x === "object" && PROJECT_SYMBOL in x;
}

export function isComponent(x: any): x is Component {
  return x !== null && typeof x === "object" && COMPONENT_SYMBOL in x;
}
