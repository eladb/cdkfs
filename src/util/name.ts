import { Project } from "../project";

/**
 * Generate workflow name with suffix based on if project is subproject or not
 * @param base name prefix
 * @param project to which the workflow belongs
 */
export function workflowNameForProject(base: string, project: Project): string {
  // Subprojects
  if (project.parent) {
    return `${base}_${fileSafeName(project.name)}`;
  }

  // root project doesn't get a suffix
  return base;
}

function fileSafeName(name: string): string {
  return name.replace("@", "").replace(/\//g, "-");
}
