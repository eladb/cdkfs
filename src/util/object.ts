/**
 *
 * @param obj Any object
 * @returns The same object but with no properties whose value is `null` or `undefined`
 */
export const removeNullOrUndefinedProperties = <
  TObj extends Record<string, any>
>(
  obj: TObj
): TObj => {
  return Object.fromEntries(
    Object.entries(obj).filter(
      ([_, value]) => value !== undefined && value !== null
    )
  ) as TObj;
};
