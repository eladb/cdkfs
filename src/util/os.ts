/**
 * Provides information about the current operating system.
 */
export class OsInspector {
  /**
   * Whether the current OS is AIX.
   */
  public readonly isAix: boolean;

  /**
   * Whether the current OS is FreeBSD.
   */
  public readonly isFreebsd: boolean;

  /**
   * Whether the current OS is Linux.
   */
  public readonly isLinux: boolean;

  /**
   * Whether the current OS is MacOS.
   */
  public readonly isMacos: boolean;

  /**
   * Whether the current OS is OpenBSD.
   */
  public readonly isOpenbsd: boolean;

  /**
   * Whether the current OS is SunOS.
   */
  public readonly isSunos: boolean;

  /**
   * Whether the current OS is Windows.
   */
  public readonly isWindows: boolean;

  /**
   * Contains a list of file extensions that the current Windows installation considers to be executable files
   */
  private _pathExt?: Set<string>;

  constructor() {
    this.isAix = process.platform === "aix";
    this.isWindows = process.platform === "win32";
    this.isMacos = process.platform === "darwin";
    this.isFreebsd = process.platform === "freebsd";
    this.isLinux = process.platform === "linux";
    this.isOpenbsd = process.platform === "openbsd";
    this.isSunos = process.platform === "sunos";
  }

  public get pathExt(): Set<string> {
    if (!this.isWindows) {
      throw new Error("pathExt is only available on Windows");
    }

    if (!this._pathExt) {
      this._pathExt = new Set(
        process.env.PATHEXT?.split(";").map((extension) =>
          extension.toLowerCase()
        )
      );
    }

    return this._pathExt;
  }
}
