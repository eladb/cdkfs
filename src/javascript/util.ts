import { existsSync, readFileSync } from "fs";
import { basename, dirname, extname, join, sep, resolve } from "path";
import * as semver from "semver";
import { findUp } from "../util";

/**
 * Basic interface for `package.json`.
 */
interface PackageManifest {
  [key: string]: any;

  /**
   * Package name.
   */
  name: string;
  /**
   * Package version.
   */
  version: string;
}

export function renderBundleName(entrypoint: string) {
  const parts = join(entrypoint).split(sep);
  if (parts[0] === "src") {
    parts.shift(); // just remove 'src' if its the first element for ergonomics
  }

  const p = parts.join(sep);
  const dir = dirname(p);
  const base = basename(p, extname(p));
  return join(dir, base);
}

/**
 * Regex for AWS CodeArtifact registry
 */
export const codeArtifactRegex =
  /^https:\/\/(?<registry>(?<domain>[^\.]+)-(?<accountId>\d{12})\.d\.codeartifact\.(?<region>[^\.]+).*\.amazonaws\.com\/.*\/(?<repository>[^\/.]+)\/)/;

/**
 * gets AWS details from the Code Artifact registry URL
 * throws exception if not matching expected pattern
 * @param registryUrl Code Artifact registry URL
 * @returns object containing the (domain, accountId, region, repository)
 */
export function extractCodeArtifactDetails(registryUrl: string) {
  const match = registryUrl.match(codeArtifactRegex);
  if (match?.groups) {
    const { domain, accountId, region, repository, registry } = match.groups;
    return { domain, accountId, region, repository, registry };
  }
  throw new Error("Could not get CodeArtifact details from npm Registry");
}

export function minVersion(version: string): string | undefined {
  if (semver.validRange(version)) {
    return semver.minVersion(version)?.version;
  } else {
    return version;
  }
}

/**
 * Attempt to resolve location of the given `moduleId`.
 * @param moduleId Module ID to lookup.
 * @param options Passed through to `require.resolve`.
 */
export function tryResolveModule(
  moduleId: string,
  options?: { paths: string[] }
): string | undefined {
  try {
    return require.resolve(moduleId, options);
  } catch {
    return undefined;
  }
}

/**
 * Attempt to resolve a module's manifest (package.json) path via `require.resolve` lookup.
 *
 * @remarks
 * If the target package has `exports` that differ from the default
 * (i.e, it defines the `exports` field in its manifest) and does not
 * explicitly include an entry for `package.json`, this strategy will fail.
 * See {@link tryResolveManifestPathFromDefaultExport} as an alternative.
 *
 * @param moduleId Module ID to lookup.
 * @param options Passed through to `require.resolve`.
 */
export function tryResolveModuleManifestPath(
  moduleId: string,
  options?: { paths: string[] }
): string | undefined {
  // cannot just `require('dependency/package.json')` here because
  // `options.paths` may not overlap with this node proc's resolution paths.
  const manifestId = `${moduleId}/package.json`;
  return tryResolveModule(manifestId, options);
}

/**
 * Attempt to resolve a module's manifest (package.json) path by looking for the nearest
 * `package.json` file that is an ancestor to the module's default export location.
 *
 * @param moduleId Module ID to lookup.
 * @param options Passed through to `require.resolve`.
 */
export function tryResolveManifestPathFromDefaultExport(
  moduleId: string,
  options?: { paths: string[] }
): string | undefined {
  const defaultExportPath = tryResolveModule(moduleId, options);
  if (!defaultExportPath) {
    return undefined;
  }
  const moduleDir = findUp("package.json", defaultExportPath);
  if (!moduleDir) {
    return undefined;
  }
  return join(moduleDir, "package.json");
}

/**
 * Attempt to resolve a module's manifest (package.json) path by checking for its existence under `node_modules` relative to `basePath`.
 *
 * @remarks
 * This strategy can be helpful in the scenario that a module defines
 * custom exports without `package.json` and no default export (i.e, some type definition packages).
 *
 * @param moduleId Module ID to lookup.
 * @param basePath Root path to search from.
 */
export function tryResolveManifestPathFromPath(
  moduleId: string,
  basePath: string
) {
  const base = basePath.includes("node_modules")
    ? basePath
    : join(basePath, "node_modules");
  const filePath = resolve(base, ...moduleId.split("/"), "package.json");
  if (existsSync(filePath)) {
    return filePath;
  }
  return undefined;
}

/**
 * Attempt to resolve a module's manifest (package.json) path by searching for it in the optionally provided paths array
 * as well as the current node processes' default resolution paths.
 * @param moduleId Module ID to search for.
 * @param options Search options.
 */
export function tryResolveManifestPathFromSearch(
  moduleId: string,
  options?: { paths: string[] }
): string | undefined {
  const searchPaths = [
    ...(options?.paths ?? []),
    ...(require.resolve.paths(moduleId) ?? []),
  ];
  for (const path of searchPaths) {
    const result = tryResolveManifestPathFromPath(moduleId, path);
    // early return on first result.
    if (result) {
      return result;
    }
  }
  return undefined;
}

/**
 * Attempt to resolve a module's manifest (package.json) using multiple strategies.
 * @param moduleId Module to resolve manifest path for.
 * @param options Resolution options.
 */
export function tryResolveModuleManifest(
  moduleId: string,
  options?: { paths: string[] }
): PackageManifest | undefined {
  const strategies = [
    tryResolveModuleManifestPath,
    tryResolveManifestPathFromDefaultExport,
    tryResolveManifestPathFromSearch,
  ];
  for (const strategy of strategies) {
    const result = strategy(moduleId, options);
    // early return on first result.
    if (result) {
      try {
        const manifest = JSON.parse(
          readFileSync(result, "utf8")
        ) as PackageManifest;
        // verify name matches target module.
        if (manifest.name === moduleId) {
          return manifest;
        }
      } catch {
        // continue to next strategy.
      }
    }
  }
  return undefined;
}

/**
 * Attempt to resolve the installed version of a given dependency.
 * @param dependencyName Name of dependency.
 * @param options Optional options passed through to `require.resolve`.
 */
export function tryResolveDependencyVersion(
  dependencyName: string,
  options?: { paths: string[] }
): string | undefined {
  const manifest = tryResolveModuleManifest(dependencyName, options);
  if (!manifest) {
    return undefined;
  }
  return manifest?.version;
}

/**
 * Whether the given requestedRange *probably* leads to the installation of a version that matches checkRange
 *
 * We assume that NPM always installs the most recent version of a package that
 * is allowed by the requestedRange.
 */
export function installedVersionProbablyMatches(
  requestedRange: string,
  checkRange: string
): boolean {
  const options = {
    includePrerelease: true,
    loose: true,
  };

  // No questions asked: always true
  if (semver.subset(requestedRange, checkRange, options)) {
    return true;
  }

  // Also no questions asked: always false
  if (!semver.intersects(requestedRange, checkRange, options)) {
    return false;
  }

  // Now we're in tricky territory. We intersect, but aren't a full subset.
  // We are in one of the following 2 situations, which we will tie-break by
  // assuming NPM will install the most recent matching version in 'requested'.
  //
  // requested  | check    | result
  // -----------|----------|-----------
  //   >= 2     |  >= 3    | probably true (chance of FP)
  //   <= 2     |  <= 1    | probably false (change of FN)
  //
  // `semver` doesn't make it easy to distinguish these cases (we can't request
  // the `maxVersion` that satisfies a range). Instead what we do is
  // get the `minVersion` of each range, and if they compare equal we assume
  // we're in the bottom case with `<=` checks, and return `false`.

  return !semver.eq(
    semver.minVersion(requestedRange, options) ?? "1.2.3",
    semver.minVersion(checkRange, options) ?? "1.2.3"
  );
}
