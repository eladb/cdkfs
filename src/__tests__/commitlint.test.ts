import { TypeScriptProject } from '..';
import { LogLevel } from '../logger';
import { mkdtemp, synthSnapshot } from './util';

test('commit lint config', () => {
  // GIVEN
  const project = new TypeScriptProject({
    outdir: mkdtemp(),
    name: 'test',
    logging: { level: LogLevel.OFF },
    defaultReleaseBranch: 'main',
    commitlint: true,
  });

  // THEN
  expect(synthSnapshot(project)['commitlint.config.js']).toMatchSnapshot();
  expect(synthSnapshot(project)['.husky/commit-msg']).toMatchSnapshot();
});