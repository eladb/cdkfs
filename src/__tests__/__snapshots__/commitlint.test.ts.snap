// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`commit lint config 1`] = `"module.exports = {extends: ['@commitlint/config-conventional']};"`;

exports[`commit lint config 2`] = `undefined`;
